File: app/src/main/java/com/kent/university/privelt/ui/risk_value/BarActivity.java
package com.kent.university.privelt.ui.risk_value;

import android.graphics.Color;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModelProviders;

import com.github.mikephil.charting.charts.BarChart;
import com.github.mikephil.charting.components.Legend;
import com.github.mikephil.charting.components.XAxis;
import com.github.mikephil.charting.components.XAxis.XAxisPosition;
import com.github.mikephil.charting.data.BarData;
import com.github.mikephil.charting.data.BarDataSet;
import com.github.mikephil.charting.data.BarEntry;
import com.github.mikephil.charting.formatter.LargeValueFormatter;
import com.github.mikephil.charting.interfaces.datasets.IBarDataSet;
import com.kent.university.privelt.R;
import com.kent.university.privelt.base.BaseActivity;
import com.kent.university.privelt.injections.Injection;
import com.kent.university.privelt.injections.ViewModelFactory;
import com.kent.university.privelt.model.Service;
import com.kent.university.privelt.model.UserData;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Random;
import java.util.Set;

import static com.kent.university.privelt.ui.risk_value.RiskValueActivity.PARAM_DATA;
import static com.kent.university.privelt.ui.risk_value.RiskValueActivity.PARAM_SERVICE;
import static com.kent.university.privelt.utils.sentence.SentenceAdapter.capitaliseFirstLetter;

public class BarActivity extends BaseActivity {

    private BarChart chart;

    private RiskValueViewModel riskValueViewModel;

    private List<Service> services;
    private List<UserData> userDatas;

    private String type;
    private String service;
    private boolean isDataCentric;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_barchart);

        services = new ArrayList<>();
        userDatas = new ArrayList<>();

        if (savedInstanceState != null) {
            type = savedInstanceState.getString(PARAM_DATA);
            service = savedInstanceState.getString(PARAM_SERVICE);
        }
        else if (getIntent() != null) {
            type = getIntent().getStringExtra(PARAM_DATA);
            service = getIntent().getStringExtra(PARAM_SERVICE);
        }

        isDataCentric = !(this.type == null || this.type.isEmpty());

        if (isDataCentric)
            setTitle(capitaliseFirstLetter(type));
        else
            setTitle(service);

        chart = findViewById(R.id.chart1);

        chart.getDescription().setEnabled(false);

        // scaling can now only be done on x- and y-axis separately
        chart.setPinchZoom(false);

        chart.setDrawBarShadow(false);
        chart.setDrawGridBackground(false);

        XAxis xAxis = chart.getXAxis();
        xAxis.setPosition(XAxisPosition.BOTTOM);
        xAxis.setDrawGridLines(false);

        chart.getAxisLeft().setDrawGridLines(false);

        // add a nice and smooth animation
        chart.animateY(1500);

        Legend l = chart.getLegend();
        l.setVerticalAlignment(Legend.LegendVerticalAlignment.BOTTOM);
        l.setHorizontalAlignment(Legend.LegendHorizontalAlignment.LEFT);
        l.setOrientation(Legend.LegendOrientation.HORIZONTAL);
        l.setDrawInside(false);
        l.setForm(Legend.LegendForm.SQUARE);
        l.setFormSize(9f);
        l.setTextSize(11f);
        l.setXEntrySpace(4f);

        configureViewModel();

        getServices();

        getUserdatas();
    }

    @Override
    protected void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);

        outState.putString(PARAM_SERVICE, service);
        outState.putString(PARAM_DATA, type);
    }

    private void configureViewModel() {
        ViewModelFactory viewModelFactory = Injection.provideViewModelFactory(this);
        riskValueViewModel = ViewModelProviders.of(this, viewModelFactory).get(RiskValueViewModel.class);
        riskValueViewModel.init();
    }

    private void updateUserDatas(List<UserData> userData) {
        this.userDatas = userData;
        if (!userDatas.isEmpty() && !services.isEmpty())
            loadData();
    }

    private void updateServices(List<Service> services) {
        this.services = services;
        if (!userDatas.isEmpty() && !services.isEmpty())
            loadData();
    }

    //TODO: refactor les requêtes SQL
    private void getServices() {
        riskValueViewModel.getServices().observe(this, this::updateServices);
    }

    private void getUserdatas() {
        riskValueViewModel.getUserDatas().observe(this, this::updateUserDatas);
    }

    private String[] getNumberOfTypes() {
        Set<String> types = new HashSet<>();
        for (UserData userData : userDatas) {
            types.add(userData.getType());
        }
        return types.toArray(new String[types.size()]);
    }

    private int countTypeForEachService(String mActivity, Service service) {
        int count = 0;
        for (UserData userData : userDatas) {
            if (userData.getServiceId() == service.getId() && userData.getType().equals(mActivity))
                count += 1;
        }
        return count;
    }

    private int countServiceForEachType(String mActivity, Service service) {
        int count = 0;
        for (UserData userData : userDatas) {
            if (userData.getServiceId() == service.getId() && userData.getType().equals(mActivity) && mActivity.equals(userData.getType()))
                count += 1;
        }
        return count;
    }

    public void loadData() {

        List<ArrayList<BarEntry>> allValues = new ArrayList<>();

        String[] mActivities = getNumberOfTypes();

        int i = 0;
        if (!isDataCentric) {
            for (Service service : services) {
                if (service.getName().equals(this.service)) {
                    for (String mActivity : mActivities) {
                        int val = countTypeForEachService(mActivity, service);
                        ArrayList<BarEntry> values = new ArrayList<>();
                        values.add(new BarEntry(i++, val));
                        allValues.add(values);
                    }
                    break;
                }
            }
        }
        else {
            for (String type : mActivities) {
                if (type.equals(this.type)) {
                    for (Service service : services) {
                        int val = countServiceForEachType(type, service);
                        ArrayList<BarEntry> values = new ArrayList<>();
                        values.add(new BarEntry(i++, val));
                        allValues.add(values);
                    }
                    break;
                }
            }
        }

        List<IBarDataSet> sets = new ArrayList<>();

        int j = 0;
        for (ArrayList<BarEntry> values : allValues) {
            BarDataSet set = new BarDataSet(values, isDataCentric ? services.get(j).getName() : mActivities[j]);
            Random rnd = new Random();
            int color = Color.argb(255, rnd.nextInt(256), rnd.nextInt(256), rnd.nextInt(256));
            set.setColor(color);
            sets.add(set);
            j++;
        }

        BarData data = new BarData(sets);
        data.setValueFormatter(new LargeValueFormatter());
        chart.setData(data);

        float groupSpace = 0.08f;
        float barSpace = 0.03f;
        float barWidth = 0.4f;

        chart.getBarData().setBarWidth(barWidth);

        data.getGroupWidth(groupSpace, barSpace);
        chart.getXAxis().setAxisMinimum(0);
        chart.getXAxis().setAxisMaximum(allValues.size() / 2);
        if (allValues.size() > 1)
            chart.groupBars(0, groupSpace, barSpace);
        chart.invalidate();
        chart.setFitBars(true);

        chart.invalidate();
    }
}

File: app/src/main/java/com/kent/university/privelt/ui/risk_value/BarActivity.kt
/*
 *  This Source Code Form is subject to the terms of the Mozilla Public
 *  * License, v. 2.0. If a copy of the MPL was not distributed with this
 *  * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package com.kent.university.privelt.ui.risk_value

import android.graphics.Color
import android.os.Bundle
import androidx.lifecycle.Observer
import com.github.mikephil.charting.components.Legend
import com.github.mikephil.charting.components.XAxis.XAxisPosition
import com.github.mikephil.charting.data.BarData
import com.github.mikephil.charting.data.BarDataSet
import com.github.mikephil.charting.data.BarEntry
import com.github.mikephil.charting.formatter.LargeValueFormatter
import com.github.mikephil.charting.interfaces.datasets.IBarDataSet
import com.kent.university.privelt.R
import com.kent.university.privelt.base.BaseActivity
import com.kent.university.privelt.model.Service
import com.kent.university.privelt.model.UserData
import com.kent.university.privelt.utils.sentence.SentenceAdapter
import kotlinx.android.synthetic.main.activity_barchart.*
import java.util.*

class BarActivity : BaseActivity() {
    private var riskValueViewModel: RiskValueViewModel? = null
    private var services: List<Service>? = null
    private var userDatas: List<UserData>? = null
    private var type: String? = null
    private var service: String? = null
    private var isDataCentric = false
    override fun configureDesign(savedInstanceState: Bundle?) {
        services = ArrayList()
        userDatas = ArrayList()
        if (savedInstanceState != null) {
            type = savedInstanceState.getString(RiskValueActivity.PARAM_DATA)
            service = savedInstanceState.getString(RiskValueActivity.PARAM_SERVICE)
        } else if (intent != null) {
            type = intent.getStringExtra(RiskValueActivity.PARAM_DATA)
            service = intent.getStringExtra(RiskValueActivity.PARAM_SERVICE)
        }
        isDataCentric = !(type == null || type!!.isEmpty())
        title = if (isDataCentric) SentenceAdapter.capitaliseFirstLetter(type) else service
        chart!!.description.isEnabled = false

        // scaling can now only be done on x- and y-axis separately
        chart!!.setPinchZoom(false)
        chart!!.setDrawBarShadow(false)
        chart!!.setDrawGridBackground(false)
        val xAxis = chart!!.xAxis
        xAxis.position = XAxisPosition.BOTTOM
        xAxis.setDrawGridLines(false)
        chart!!.axisLeft.setDrawGridLines(false)

        // add a nice and smooth animation
        chart!!.animateY(1500)
        val l = chart!!.legend
        l.verticalAlignment = Legend.LegendVerticalAlignment.BOTTOM
        l.horizontalAlignment = Legend.LegendHorizontalAlignment.LEFT
        l.orientation = Legend.LegendOrientation.HORIZONTAL
        l.setDrawInside(false)
        l.form = Legend.LegendForm.SQUARE
        l.formSize = 9f
        l.textSize = 11f
        l.xEntrySpace = 4f
        configureViewModel()
        getServices()
        userdatas
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putString(RiskValueActivity.PARAM_SERVICE, service)
        outState.putString(RiskValueActivity.PARAM_DATA, type)
    }

    override val activityLayout: Int
        get() = R.layout.activity_barchart

    override fun configureViewModel() {
        riskValueViewModel = getViewModel(RiskValueViewModel::class.java)
        riskValueViewModel?.init()
    }

    private fun updateUserDatas(userData: List<UserData>) {
        userDatas = userData
        if (userDatas!!.isNotEmpty() && services!!.isNotEmpty()) loadData()
    }

    private fun updateServices(services: List<Service>) {
        this.services = services
        if (userDatas!!.isNotEmpty() && services.isNotEmpty()) loadData()
    }

    //TODO: refactor les requêtes SQL
    private fun getServices() {
        riskValueViewModel!!.services?.observe(this, Observer { services: List<Service> -> updateServices(services) })
    }

    private val userdatas: Unit
        get() {
            riskValueViewModel!!.userDatas?.observe(this, Observer { userData: List<UserData> -> updateUserDatas(userData) })
        }

    private val numberOfTypes: Array<String>
        get() {
            val types: MutableSet<String> = HashSet()
            for ((_, type1) in userDatas!!) {
                types.add(type1)
            }
            return types.toTypedArray()
        }

    private fun countTypeForEachService(mActivity: String, service: Service): Float {
        var count = 0F
        for ((_, type1, _, _, serviceId) in userDatas!!) {
            if (serviceId == service.id && type1 == mActivity) count += 1
        }
        return count
    }

    private fun countServiceForEachType(mActivity: String, service: Service): Float {
        var count = 0F
        for ((_, type1, _, _, serviceId) in userDatas!!) {
            if (serviceId == service.id && type1 == mActivity && mActivity == type1) count += 1
        }
        return count
    }

    private fun loadData() {
        val allValues: MutableList<ArrayList<BarEntry>> = ArrayList()
        val mActivities = numberOfTypes
        var i = 0F
        if (!isDataCentric) {
            for (service in services!!) {
                if (service.name == this.service) {
                    for (mActivity in mActivities) {
                        val value = countTypeForEachService(mActivity, service)
                        val values = ArrayList<BarEntry>()
                        values.add(BarEntry(i++, value))
                        allValues.add(values)
                    }
                    break
                }
            }
        } else {
            for (type in mActivities) {
                if (type == this.type) {
                    for (service in services!!) {
                        val value = countServiceForEachType(type, service)
                        val values = ArrayList<BarEntry>()
                        values.add(BarEntry(i++, value))
                        allValues.add(values)
                    }
                    break
                }
            }
        }
        val sets: MutableList<IBarDataSet> = ArrayList()
        for ((j, values) in allValues.withIndex()) {
            val set = BarDataSet(values, if (isDataCentric) services!![j].name else mActivities[j])
            val rnd = Random()
            val color = Color.argb(255, rnd.nextInt(256), rnd.nextInt(256), rnd.nextInt(256))
            set.color = color
            sets.add(set)
        }
        val data = BarData(sets)
        data.setValueFormatter(LargeValueFormatter())
        chart!!.data = data
        val groupSpace = 0.08f
        val barSpace = 0.03f
        val barWidth = 0.4f
        chart!!.barData.barWidth = barWidth
        data.getGroupWidth(groupSpace, barSpace)
        chart!!.xAxis.axisMinimum = 0f
        chart!!.xAxis.axisMaximum = allValues.size / 2.toFloat()
        if (allValues.size > 1) chart!!.groupBars(0f, groupSpace, barSpace)
        chart!!.invalidate()
        chart!!.setFitBars(true)
        chart!!.invalidate()
    }
}